extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js

```diff
@@ -65,17 +69,117 @@ describe('amp-ad-api-handler', () => {
       }));
     });
 
-    it('should resolve startUp() when render-start API is called', () => {
-      expect(iframe.style.visibility).to.equal('hidden');
-      iframe.postMessageToParent({
-        sentinel: 'amp3ptest' + testIndex,
-        type: 'render-start',
+    describe('method startUp return promise', () => {
+      it('should resolve on message "bootstrap-loaded" if render-start is'
+          + 'NOT implemented by 3P', () => {
+        expect(iframe.style.visibility).to.equal('hidden');
+        iframe.postMessageToParent({
+          sentinel: 'amp3ptest' + testIndex,
+          type: 'bootstrap-loaded',
+        });
+        return startUpPromise.then(() => {
+          expect(iframe.style.visibility).to.equal('');
+        });
+      });
+
+      it('should resolve on message "render-start" if render-start is'
+          + 'implemented by 3P"', () => {
+        adImpl.adType = 'doubleclick';
+        apiHandler = new AmpAdApiHandler(adImpl, adImpl.element, () => {});
+        const beforeAttachedToDom = element => {
+          element.setAttribute('data-amp-3p-sentinel', 'amp3ptest' + testIndex);
+          startUpPromise = apiHandler.startUp(element, true);
+        };
+        const noContentCallbackSpy
+            = sandbox.spy(apiHandler, 'noContentCallback_');
+        return createIframeWithMessageStub(window, beforeAttachedToDom)
+            .then(newIframe => {
+              iframe = newIframe;
+              expect(iframe.style.visibility).to.equal('hidden');
+              iframe.postMessageToParent({
+                sentinel: 'amp3ptest' + testIndex,
+                type: 'render-start',
+              });
+              return startUpPromise.then(() => {
+                expect(iframe.style.visibility).to.equal('');
+              }).then(() => {
+                iframe.postMessageToParent({
+                  sentinel: 'amp3ptest' + testIndex,
+                  type: 'no-content',
+                });
+                return expectPostMessage(iframe.contentWindow, window, {
+                  sentinel: 'amp3ptest' + testIndex,
+                  type: 'no-content',
+                }).then(() => {
+                  expect(noContentCallbackSpy).to.not.been.called;
+                });
+              });
+            });
       });
-      return startUpPromise.then(() => {
-        expect(iframe.style.visibility).to.equal('');
+
+      it('should resolve on message "no-content" if render-start is'
+          + 'implemented by 3P', () => {
+        adImpl.adType = 'doubleclick';
+        apiHandler = new AmpAdApiHandler(adImpl, adImpl.element, () => {});
+        const beforeAttachedToDom = element => {
+          element.setAttribute('data-amp-3p-sentinel', 'amp3ptest' + testIndex);
+          startUpPromise = apiHandler.startUp(element, true);
+        };
+        const noContentCallbackSpy
```

Comment:
if you want a spy, you can just create one and pass it into AmpAdApiHandler's constructor.
try to prevent access private member directly.